src/enftun_xtt.c

```diff
@@ -0,0 +1,894 @@
+        rc = xtt_initialize_client_group_context_lrsw(group_ctx,
+                                             &gid,
+                                             &daa_priv_key,
+                                             &cred,
+                                             (unsigned char*)basename,
+                                             basename_len);
+    }
+
+    if (XTT_RETURN_SUCCESS != rc){
+        printf("%s", xtt_strerror(rc));
+        return TPM_ERROR;
+    }
+
+    return 0;
+}
+
+static
+int initialize_certs(int use_tpm,
+                     TSS2_TCTI_CONTEXT *tcti_context, xtt_root_certificate* root_certificate)
+{
+    xtt_return_code_type rc = 0;
+    // 1) Read root id ang pubkey in from buffer
+    xtt_certificate_root_id root_id = {.data = {0}};
+    xtt_ecdsap256_pub_key root_public_key = {.data = {0}};
+
+    if (use_tpm) {
+#ifdef USE_TPM
+        int nvram_ret;
+        nvram_ret = read_nvram(root_id.data,
+                               sizeof(xtt_certificate_root_id),
+                               root_id_handle_g,
+                               tcti_context);
+        if (0 != nvram_ret) {
+            fprintf(stderr, "Error reading root ID from TPM NVRAM");
+            return TPM_ERROR;
+        }
+
+        nvram_ret = read_nvram(root_public_key.data,
+                               sizeof(xtt_ecdsap256_pub_key),
+                               root_pubkey_handle_g,
+                               tcti_context);
+        if (0 != nvram_ret) {
+            fprintf(stderr, "Error reading root's public key from TPM NVRAM");
+            return TPM_ERROR;
+        }
+#else
+        fprintf(stderr, "Attempted to use a TPM, but not built with TPM enabled!\n");
+        return TPM_ERROR;
+#endif
+    } else {
+        memcpy(root_public_key.data, &root_certificate->data[sizeof(xtt_identity_type)], sizeof(xtt_ecdsap256_pub_key));
+        memcpy(root_id.data, &root_certificate->data[0], sizeof(xtt_certificate_root_id));
+    }
+
+    // 2) Initialize stored data
+    memcpy(stored_root_id, root_id.data, sizeof(xtt_certificate_root_id));
+
+    rc = xtt_initialize_server_root_certificate_context_ecdsap256(&stored_cert,
+                                                                &root_id,
+                                                                &root_public_key);
+    if (XTT_RETURN_SUCCESS != rc)
+        return CLIENT_ERROR;
+
+    return 0;
+}
+
+static
+int do_handshake_client(int socket,
+                 xtt_identity_type *requested_client_id,
+                 xtt_identity_type *intended_server_id,
+                 struct xtt_client_group_context *group_ctx,
+                 struct xtt_client_handshake_context *ctx)
+{
+    xtt_return_code_type rc = XTT_RETURN_SUCCESS;
+
+    uint16_t bytes_requested = 0;
+    unsigned char *io_ptr = NULL;
+    xtt_certificate_root_id claimed_root_id = {.data = {0}};
+    printf("Starting identity-provisioning handshake, by sending ClientInit message...\n");
```

Comment:
You can remove all of these diagnostic info messages.

Just keep (some of) the ones for errors.